Support DPI change in sizers

MaartenBent opened the pull request:
Return the size of `DoGetDefaultBorderInPx` as double, so no precision is lost when multiplying it for `DoubleBorder` and `TripleBorder`.

Closes [#18551](https://trac.wxwidgets.org/ticket/18551).

PBfordev commented on include/wx/sizer.h:
```diff
@@ -223,7 +228,7 @@ class WXDLLIMPEXP_CORE wxSizerFlags
 private:
 #ifdef wxNEEDS_BORDER_IN_PX
-    static int DoGetDefaultBorderInPx();
+    static double DoGetDefaultBorderInPx();
```
Just curious, you use `double` here and in `wxPrivate::DpiDependentValue<double>` while `float` seems to be used elsewhere for scaling with DPI?

MaartenBent replied:
`GetContentScaleFactor()` returns double so I kept using that. I can use float if this is preferred.

PBfordev replied:
I see, sorry for the noise.

MaartenBent replied:
It did make me look at the code again, and I noticed `C6011: Dereferencing NULL pointer 'win'.` IntelliSense warning. Every remark can be useful.

PBfordev commented on src/msw/window.cpp:
```diff
@@ -4884,6 +4884,38 @@ wxWindowMSW::MSWUpdateOnDPIChange(const wxSize& oldDPI, const wxSize& newDPI)
+        {
+            wxSizerItem* sizerItem = current->GetData();
+
+            int borderOld = sizerItem->GetBorder();
```
Is `borderOld` really needed?

MaartenBent replied:
No, its a remnant from my debugging output.

PBfordev reviewed (commented):
Thanks again.

vadz commented on include/wx/sizer.h:
```diff
@@ -106,6 +106,11 @@ class WXDLLIMPEXP_CORE wxSizerFlags
+        return wxRound(GetDefaultBorderDouble());
+    }
+
+    static double GetDefaultBorderDouble()
```
It's a nitpick but I wonder if we can use something better than "Double" in the new function name. Unfortunately I don't have any great ideas myself, something like `GetDefaultBorderFractionalPx()` is probably too long, but maybe still worth it?

Another idea would be to add `static int GetScaledDefaultBorder(int multiple)` that would avoid using `double` in the interface completely and use them only internally.

MaartenBent replied:
I will rename it. It can also cause confusion with `DoubleBorder()` where double has a different meaning.

vadz commented on src/msw/window.cpp:
```diff
@@ -4884,6 +4884,38 @@ wxWindowMSW::MSWUpdateOnDPIChange(const wxSize& oldDPI, const wxSize& newDPI)
+    {
+        wxSizerItemList::compatibility_iterator current =
+            GetSizer()->GetChildren().GetFirst();
+        while ( current )
```
I prefer writing these loops as
```
        for ( wxWindowList::compatibility_iterator node = GetChildren().GetFirst();
              node;
              node = node->GetNext() )
```
but it's pretty minor, of course. A `for` loop would be simpler to replace with ranged for in the future...

vadz reviewed (commented):
Thanks!

As usual, I'm fine with applying this as is, except that the new function needs to be documented.

But please let me know if you see any merit in either of my proposals.

MaartenBent committed:
Support DPI change in sizers

Return the size of DoGetDefaultBorderInPx as float, so no precision is lost
when multiplying it for DoubleBorder and TripleBorder.

Closes #18551.

MaartenBent committed:
Use for-loop to iterate sizer and window children

vadz commented:
Thanks again!
